Fix float to int conversion bugs in JBMC

peterschrammel opened the pull request:
<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [ ] Each commit message has a non-empty body, explaining why the change was made.
- [X] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [X] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [X] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

peterschrammel committed:
Add tests that exhibit float to int conversion bugs in JBMC

tautschnig commented:
@peterschrammel Needs a rebase to make CI pass.

martin-cs commented on jbmc/regression/jbmc/FloatToIntConversion/Test.java:
```diff
@@ -0,0 +1,220 @@
+        // use constant here, -0.0f reverses cases 2 and 3.
+        int result = (int)(a / 0.0f);
+        // a/b = +inf
+        assert result == 2147483647;
```
I'm going to trust you that this is the JVM semantics but ... I am surprised by this.

peterschrammel replied:
I was also surprised until I ran it through the JVM.

martin-cs replied:
Cool.

martin-cs commented on jbmc/regression/jbmc/FloatToIntConversion/Test.java:
```diff
@@ -0,0 +1,220 @@
+      }
+      break;
+    case 4: 
+      if (a == 2147483646.0f && b == 1.0f) {
```
There is quite a lot going on here and I feel like, without the comment it is easy to miss some of it.  Here is my understanding:

`2147483646` is `0x7FFFFFFE` which has 30 significant bits so `2147483646.0f` will round /in the compiler/ to `0x1p+31f` a.k.a. 2^31 which is the same as `2147483648.0f`.  The divide is effectively a no-op and so this tests the conversion to `int` but also the compiler's handling of constants.

( There are some really horrific examples like this where you can write code that will, at run-time, tell you whether the compiler correctly parsed the number as a float or whether it parsed it first as a double and cast down. )

peterschrammel replied:
There are quite a lot of subtle things going on. I can comment in more detail what is happening in the test, and maybe choose constants that are not going to be rounded. I should probably also explain the effect of +1 (resp. -1) in java_bytecode_convert_method.cpp:1681, which are subject to similar rounding effects (i.e. it may not be totally obvious why this works).

martin-cs replied:
With this kind of thing, more documentation is better than less.  I know you get it and intend what happens but it should be commented so that others can learn that.

martin-cs commented on jbmc/regression/jbmc/FloatToIntConversion/Test.java:
```diff
@@ -0,0 +1,220 @@
+      }
+      break;
+    case 4: 
+      if (a == 2147483649.0f && b == 1.0f) {
```
A similar thing happens here.

peterschrammel replied:
@martin-cs, improved.

martin-cs replied:
Thanks.

martin-cs reviewed (approved):
Looks good.  Some more comments on some of the rounding effects would be helpful IMHO.

peterschrammel committed:
Fix float to int conversion bugs in JBMC

Conversion have to follow JLS 5.1.3. Narrowing Primitive Conversion

Roughly speaking, non-representable numbers saturate
at INT_MIN and INT_MAX. NaN is converted to 0.
